File: 2021/07/09/main.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = []
        for i in range(len(nums)):
            idx = binary_search(dp, nums[i])
            k = len(dp)
            if idx == k:
                arr = []
                if k != 0:
                    arr = [i for i in dp[-1]]
                arr.append(nums[i])
                dp.append(arr)
            elif dp[idx][-1] > nums[i]:
                dp[idx][-1] = nums[i]

        return len(dp[-1])

def binary_search(dp, target):
    l = 0
    r = len(dp) - 1
    while l <= r:
        mid = l + (r - l) // 2
        if dp[mid][-1] == target:
            return mid
        elif dp[mid][-1] < target:
            l = mid + 1
        else:
            r = mid - 1
    return l

File: 2021/07/09/main.go
package main

func lengthOfLIS(nums []int) int {
	dp = make([][]int)
	for i := 0; i < len(nums); i++ {
		idx := bs(dp, nums[i])
		k = len(dp)
		if idx == k {
			arr = []int{}
			if k != 0 {
				for j := 0; j < dp[k-1]; j++ {
					arr = append(arr, j)
				}
			}
			arr = append(arr, nums[i])
			dp = append(dp, arr)
		} else if dp[idx][-1] > nums[i] {
			dp[idx][-1] = nums[i]
		}
	}
	return dp[-1]
}

func bs(dp []int, t int) int {
	l, r := 0, len(dp)-1
	for l <= r {
		m := l + (r-l)/2
		d := dp[m]
		if d[-1] == t {
			return m
		} else if d[-1] < t {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	return l
}
